Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,890 @@
+    /// This returns the default EventLoop that a LambdaRuntime is scheduled on.
+    /// It uses `NIOSingletons.posixEventLoopGroup.next()` under the hood.
+    public static var defaultEventLoop: any EventLoop { get }
+
+    /// Report a startup error to the Lambda Control Plane API
+    public static func reportStartupError(any Error) async
+}
+```
+
+Since the library now provides ownership of the `main()` function and allows users to initialize services before the
+`LambdaRuntime` is initialized, the library cannot implicitly report
+[errors that occur during initialization to the dedicated endpoint AWS exposes](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-api.html#runtimes-api-initerror)
+like it currently does through the `initialize()` function of `LambdaRunner` which wraps the handler's `init(...)` and
+handles any errors thrown by reporting it to the dedicated AWS endpoint.
+
+To retain support for initialization error reporting, the `Lambda.reportStartupError(any Error)` function gives users
+the option to manually report initialization errors in their closure handler. Although this should ideally happen
+implicitly like it currently does in v1, we believe this is a small compromise in comparison to the benefits gained in
+now being able to cleanly manage the lifecycles of required services in a structured concurrency manner.
+
+> Use-case:
+>
+> Assume we want to load a secret for the Lambda function from a secret vault first. If this fails, we want to report
+> the error to the control plane:
+>
+> ```swift
+> let secretVault = SecretVault()
+>
+> do {
+>    /// !!! Error thrown: secret "foo" does not exist !!!
+>    let secret = try await secretVault.getSecret("foo")
+>
+>    let runtime = LambdaRuntime { (event: Input, context: LambdaContext) in
+>        /// Lambda business logic
+>    }
+>
+>    let serviceGroup = ServiceGroup(
+>        services: [postgresClient, runtime],
+>        configuration: .init(gracefulShutdownSignals: [.sigterm]),
+>        logger: logger
+>    )
+>    try await serviceGroup.run()
+> } catch {
+>    /// Report startup error straight away to the dedicated initialization error endpoint
+>    try await Lambda.reportStartupError(error)
+> }
+> ```
+
+### Codable support
+
+The `LambdaHandler` and `LambdaWithBackgroundProcessingHandler` protocols abstract away encoding/decoding logic from the
+conformers as they are generic over custom `Event` and `Output` types. We introduce two adapters `LambdaHandlerAdapter`
+and `CodableLambdaAdapter` that implement the encoding/decoding logic and in turn allow the respective handlers to
+conform to `StreamingLambdaHandler`.
+
+#### LambdaHandlerAdapter
+
+Any handler conforming to `LambdaHandler` can be conformed to `LambdaWithBackgroundProcessingHandler` through
+`LambdaHandlerAdapter`.
+
+```swift
+/// Wraps an underlying handler conforming to ``LambdaHandler``
+/// with ``LambdaWithBackgroundProcessingHandler``.
+public struct LambdaHandlerAdapter<
```

Comment:
It's not clear how the Event is decoded and Output encoded in the `LambdaHandlerAdapter`.